apps/src/lib/node/ledger/protocol/transactions/ethereum_events/eth_msgs.rs

```diff
@@ -0,0 +1,98 @@
+use std::collections::BTreeSet;
+
+use borsh::{BorshDeserialize, BorshSchema, BorshSerialize};
+use namada::types::address::Address;
+use namada::types::ethereum_events::EthereumEvent;
+use namada::types::storage::BlockHeight;
+use namada::types::vote_extensions::ethereum_events::MultiSignedEthEvent;
+use namada::types::voting_power::FractionalVotingPower;
+
+/// Represents an Ethereum event being seen by some validators
+#[derive(
+    Debug,
+    Clone,
+    Ord,
+    PartialOrd,
+    PartialEq,
+    Eq,
+    Hash,
+    BorshSerialize,
+    BorshDeserialize,
+)]
+pub struct EthMsgUpdate {
+    /// The event being seen
+    pub body: EthereumEvent,
+    /// Addresses of the validators who have just seen this event
+    /// we use [`BTreeSet`] even though ordering is not important here, so that
+    /// we can derive [`Hash`] for [`EthMsgUpdate`]. This also conveniently
+    /// orders addresses in the order in which they should be stored in
+    /// blockchain storage.
+    // NOTE(feature = "abcipp"): This can just become BTreeSet<Address> because
+    // BlockHeight will always be the previous block
+    pub seen_by: BTreeSet<(Address, BlockHeight)>,
+}
+
+impl From<MultiSignedEthEvent> for EthMsgUpdate {
+    fn from(
+        MultiSignedEthEvent { event, signers }: MultiSignedEthEvent,
+    ) -> Self {
+        Self {
+            body: event,
+            seen_by: signers.into_iter().collect(),
+        }
+    }
+}
+
+/// Represents an event stored under `eth_msgs`
+#[derive(
+    Clone, Debug, PartialEq, Eq, BorshSerialize, BorshDeserialize, BorshSchema,
+)]
+pub struct EthMsg {
+    /// The event being stored
+    pub body: EthereumEvent,
+    /// The total voting power that's voted for this event across all epochs
+    pub voting_power: FractionalVotingPower,
+    /// The addresses of validators that voted for this event, in sorted order.
+    pub seen_by: Vec<Address>,
+    /// Whether this event has been acted on or not
+    pub seen: bool,
```

Comment:
This struct represents an event as stored under `/eth_msgs`, which has an explicit `seen` key which maps to this field. So as long as the stored event has a `seen` key, we should have this field here. Though it does seem kind of redundant in a way as it can be derived whether an event is seen or not based on the `voting_power` being >2/3 like you say. Maybe we should simplify our design and not store an explicit `seen` field at all?